Replace IPAddress implementation with managed implementation on Unix

stephentoub opened the pull request:
We have several bugs where different platforms parse IP addresses with different rules, different handling of corner cases, etc.  To address this, I'm switching the Unix implementation to be entirely managed, such that parsing strings to IPAddress and rendering IPAddress to strings is handled in managed code without any P/Invokes.

This PR:
- Ports desktop's IPv4AddressHelper and IPv6AddressHelper to core.  I made as few changes as possible to these in order to get things to work, but I did make a few changes in order to fix some bugs, e.g. IPv6AddressHelper's validation routines were letting through some invalid addresses (e.g. nothing after a closing bracket was being validated, scope ID overflow wasn't being validated, etc.)
- Replaces IPAddressPal.Unix.cs with IPAddressPal.Managed.cs.  I left in place the current PAL layer, but in the future we could easily remove the PAL entirely and just use the managed implementation (I tried compiling IPAddressPal.Managed.cs into the Windows build, and it builds and tests fine.
- The one place I did remove the PAL entirely was for IPAddressParser.IPv4AddressToString: there's no point in calling out to native code for that, so I put a managed implementation in place across the board.
- I didn't spent a lot of time on performance, but I did remove some of the aggregious things that were happening, e.g. we were allocating a new StringBuilder every time we did IPAddress.ToString, we were allocating byte[]s while parsing unnecessarily, etc.
- I consolidated some of the existing test cases into theories and added a bunch more test cases.  This revealed other bugs on Unix, which I also addressed.

cc: @geoffkizer, @Priya91, @davidsh, @CIPop, @pgavlin 
Fixes https://github.com/dotnet/corefx/issues/11843 (and some other bugs for which we don't have issues)

stephentoub commented:
(Separately, note that there's lots of opportunity for improvement.  For example, the IPv6AddressHelper ported from desktop first does a validation pass and then does a parsing pass... this should be easily combined into a single pass to improve performance.  As another example, if we remove the PAL, we can remove some unnecessary marshaling of data from one form to another that's happening, e.g. from a byte[] to a ushort[] and vice versa.  I'll open an issue for this once the change goes in.)

davidsh commented:
>We have several bugs where different platforms parse IP addresses with different rules, different handling of corner cases, etc. To address this, I'm switching the Unix implementation to be entirely managed, such that parsing strings to IPAddress and rendering IPAddress to strings is handled in managed code without any P/Invokes.

If we're doing this, then why not just have a single managed implementation? Is there any performance disadvantage to this approach in general?

stephentoub commented:
> If we're doing this, then why not just have a single managed implementation, i.e. for Windows as well?

We can.  I noted that in the PR description.  I held off on doing it as Cristian expressed concerns about it.

davidsh commented:
>We can. I noted that in the PR description. I held off on doing it as Cristian expressed concerns about it.

What are the concerns? Is it performance related? Speed or memory size? I can see if the memory footprint got higher on Windows then perhaps for UWP use of this library, it might be better to stay with native p/invokes. But our latest thinking is trending toward a single managed implementation everywhere which is easier to maintain and control behaviors.

stephentoub commented:
> Is it performance related?

I did some small measurements, and they seem on par, maybe 1-2% difference for the addresses I tried.  But as noted, there are also opportunities to clean up the ported code to do better, e.g. a single pass rather than two.  I don't think performance is a reason to not do it.  Similarly there shouldn't be any measurable memory increase: I removed the allocations that were involved, and at this point I think the only unnecessary heap allocation remaining is in a call to StringBuilder.Append for the scope ID, but that could be fixed as well (though ideally it'd be fixed in StringBuilder.Append itself).  I switched several heap allocations to be stack-based.

> But our latest thinking is trending toward a single managed implementation everywhere which is easier to maintain and control behaviors.

Great.  Then unless anyone throws up a flag, I'll delete the Windows implementation and just use the managed one everywhere.

davidsh commented:
>Great. Then unless anyone throws up a flag, I'll delete the Windows implementation and just use the managed one everywhere.

I am fine with that.

stephentoub commented:
I pushed additional commits to use the implementation everywhere, as well as a few more tweaks/perf fixes.  Across a variety of IPv4 and IPv6 addresses, throughput is basically the same as it was for ToString, and is now ~20% faster than the native implementation(s) for {Try}Parse.

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPAddressParser.cs:
```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+            }
+        }
+
+        public static unsafe bool Ipv6StringToAddress(string ipString, ushort* numbers, int numbersLength, out uint scope)
```
How and why would numbersLength ever be anything other than IPv6AddressShorts?

stephentoub replied:
In current usage it wouldn't.  It's just a safety valve to allow the caller to say how long the pointed to memory is before it's used.  I can remove if you think it's overkill.

geoffkizer commented on src/System.Net.Primitives/tests/FunctionalTests/IPAddressParsing.cs:
```diff
@@ -2,43 +2,44 @@
         [InlineData("157.3873051", "157.59.25.27")]
         [InlineData("157.6427", "157.0.25.27")]
+        [InlineData("0313.027035210", "203.92.58.136")]
+        [InlineData("0313.0134.035210", "203.92.58.136")]
```
These are octal, right?  Should they be under the octal section below?

stephentoub replied:
Yup, will move.

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPAddressParser.cs:
```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+                    ((0x00FF0000 & tmpAddr) >> 8) |
+                    ((0x0000FF00 & tmpAddr) << 8) |
+                    ((0x000000FF & tmpAddr) << 24);
+                return true;
```
Can we at least add a comment here?  I'm not sure why IPv4AddressHelper returns the bytes in the reverse order.

stephentoub replied:
I don't know why either. I'll add a comment.

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPAddressParser.cs:
```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+            {
+                buffer.Append("::");
+                needsColon = false;
+                fromInclusive = zeroEnd;
```
It's a bit odd that we assign fromInclusive here.  Why not just use zeroEnd below and avoid mutating these, for clarity?

stephentoub replied:
Use zeroEnd starting from where?

geoffkizer replied:
yeah, you are right.

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPAddressParser.cs:
```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+        /// <summary>Appends a number as hexadecimal (without the leading "0x") to the StringBuilder.</summary>
+        private static unsafe void AppendHex(int value, StringBuilder buffer)
+        {
+            const int MaxLength = 8;
```
It looks like this is always called with a short.  Is that right?  Should value be a short and MaxLength = 4?

stephentoub replied:
I'll verify and make that change if it's true (I think you're right).

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPAddressParser.cs:
```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+                char c = chars[i];
+                chars[i] = chars[len - i - 1];
+                chars[len - i - 1] = c;
+            }
```
Seems like we could just fill the buffer from the back and avoid the need to reverse.  I think that's what we do in the ipv4 case.

stephentoub replied:
We don't know how long it's going to be. If we did that, we'd either need to compute it twice or shift all the bytes down after. We don't have that problem with v4 as we copy of the data from the buffer into a string.

geoffkizer replied:
We copy here too, right?  We do buffer.Append just below.

stephentoub replied:
Ah, you're right, will fix.

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPv6AddressHelper.cs:
```diff
@@ -0,0 +1,472 @@
+
+        //  Remarks: MUST NOT be used unless all input indexes are verified and trusted.
+        //           start must be next to '[' position, or error is reported
+        internal unsafe static bool IsValidStrict(char* name, int start, ref int end)
```
I'm not clear why IsValidStrict is separate from Parse.  It looks like we only ever call these together.

geoffkizer replied:
Saw your comment above re this.

The concern I have here is that Parse assumes that IsValidStrict has validated a bunch of stuff.  It's really hard to determine whether there might be some sort of parsing error in Parse because you need to understand exactly how the input has been validated by IsValidStrict.

Maybe we should just live with this for now, but it worries me a bit...

stephentoub replied:
Yeah, it's both brittle and inefficient.  But it's what desktop has, and I think it's best to stick with it for now rather than rewriting the whole parsing routine.

geoffkizer replied:
Makes sense.

geoffkizer commented on src/System.Net.Primitives/src/System/Net/IPv6AddressHelper.cs:
```diff
@@ -0,0 +1,472 @@
+                                    && (numbers[7] == 0x0001))
+                                   && ((numbers[5] == 0)
+                                       || (numbers[5] == 0xFFFF))));
+
```
It seems like we don't actually use the return value here.  Did I miss some call site that uses this?

stephentoub replied:
No, desktop has other call sites to it that use this.  I can remove it.

geoffkizer commented:
Overall I think this looks good for 2.0.

Longer term we should revisit this code and try to clean it up.  Can we file an issue for that?

stephentoub commented:
> Overall I think this looks good for 2.0.

Thanks.

> Longer term we should revisit this code and try to clean it up. Can we file an issue for that?

Yes, I agree.

stephentoub commented:
@dotnet-bot test outerloop netcoreapp Windows_NT Debug please
@dotnet-bot test outerloop netcoreapp Ubuntu14.04 Debug please

Priya91 commented on src/System.Net.Primitives/src/System/Net/IPAddressParser.cs:
```diff
@@ -66,38 +47,217 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
-            if (errorCode == IPAddressPal.SuccessErrorCode)
+            FormatIPv4AddressNumber((int)((address >> 24) & 0xFF), addressString, ref offset);
+            addressString[--offset] = '.';
+            FormatIPv4AddressNumber((int)((address >> 16) & 0xFF), addressString, ref offset);
```
I dont understand this code, looks like the most significant address bit is put in least significant bit of the array, hence reversing the parts, but not reversing within the 8-bit parts? So, FF.00.00.01 becomes 01.00.00.FF, which means the given uint address bits are reversed?

stephentoub replied:
Right
